Esp32 fs

ChinaTizenRT opened the pull request:
this is init version, maybe nedd refactory.

ChinaTizenRT committed:
external/esp_idf_port: add ESP32 spi flash driver.

Add code for ESP32 internal 4MB spi flash.

ChinaTizenRT committed:
external/esp_idf_port: add esp32 flash access for mtd driver.

This file exports up_progmem* interfaces for on chip flash.

SRCN-LiQin commented on external/esp_idf_port/esp32/esp32_flash.c:
```diff
@@ -0,0 +1,326 @@
+	} else {
+		char *adpt_buf = (char *)kmm_malloc(count);
+		if (adpt_buf == NULL) {
+			printf("fail to alloc memory\n");
```
better use debug log instead of printf here

SRCN-LiQin commented on external/esp_idf_port/esp32/esp32_flash.c:
```diff
@@ -0,0 +1,326 @@
+	/* Restore IRQs */
+	irqrestore(irqs);
+	if (ret < 0) {
+		printf("Errno: %d, addr %d, buf %p, count %d\n", ret, addr, buf, count);
```
many printf in this file, plz check

ChinaTizenRT replied:
rewrite this file, move to arch/xtensa/src/esp32/esp32_flash.c 
and fix many printf.

ChinaTizenRT committed:
external/esp_idf_port: Add ESP32 spi flash driver dependency.

Add ESP32 SPI flash driver needed code and head files.
to make build success.

ChinaTizenRT committed:
arch/xtensa: Enable smartfs on ESP32 SPI-based flash.

1) Add ESP32 SPI-based flash MTD driver.
2) Configure partion, formate as smartfs and mount.
3) Enable partion table on hello/defconfig.

zhouxinhe commented on build/configs/esp32_DevKitC/hello_with_tash/defconfig:
```diff
@@ -156,7 +156,7 @@ CONFIG_ARCH_BOARD="esp32_devkit"
 # CONFIG_LIB_BOARDCTL is not set
 # CONFIG_BOARD_FOTA_SUPPORT is not set
-
+CONFIG_ESP32_WIFI_SUPPORT=y
```
is this file generated by menuconfig? or you add new CONFIG_* into existed file manually?
the former is suggested, otherwise, it's inconvenient to use menuconfig to modify configration.

zhouxinhe replied:
I mean, it's better to edit Kconfig files firstly and then make menuconfig to generate `.config`, at last, rename `.config` to `defconfig` and replace the old one.

zhouxinhe commented on external/esp_idf_port/esp32/system_api.c:
```diff
@@ -0,0 +1,304 @@
+/******************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
```
year 2019, also check other files.

xixidodo commented on external/esp_idf_port/spi_flash/cache_utils.c:
```diff
@@ -0,0 +1,343 @@
+	s_flash_op_cpu = cpuid;
+#endif
+
+	if (xTaskGetSchedulerState() == taskSCHEDULER_NOT_STARTED) {
```
many freertos APIs in this file, so it is not compiled?
if not, we can remove it ?

ChinaTizenRT replied:
no need to remove, this code is prepared for SMP( more cpus)
